compiler/resolution/nilChecking.cpp

```diff
@@ -0,0 +1,782 @@
+            storeToUnknown = true;
+
+        } else {
+          OUT[actualSym] = unknownAliasLocation();
+        }
+      }
+    }
+  }
+}
+
+static void checkBasicBlock(
+    FnSymbol* fn,
+    BasicBlock* bb,
+    const std::vector<Symbol*> symbols,
+    const AliasMap& IN,
+    AliasMap& OUT,
+    bool raiseErrors) {
+
+  OUT = IN;
+
+  for_vector(Expr, expr, bb->exprs) {
+
+    bool storeToUnknown = false;
+    bool storeToGlobals = false;
+
+    bool trace = (0 == strcmp(fn->name, debugNilsForFn));
+
+    AliasMap* traceCopy = NULL;
+
+    if (trace) {
+      traceCopy = new AliasMap(OUT);
+
+      printf("\nvisiting:");
+      nprint_view(expr);
+
+      printAliasMap("bb", OUT);
+    }
+
+    if (expr->id == debugNilsForId)
+      gdbShouldBreakHere();
+
+    if (isDefExpr(expr)) {
+      // Ignore DefExprs
+      // instead, find the defaultOf or whatever initializes them.
+    } else if (ForallStmt* forall = toForallStmt(expr)) {
+
+      for_alist(expr, forall->iteratedExpressions()) {
+        if (CallExpr* call = toCallExpr(expr)) {
+          checkCall(fn, bb, call, symbols, OUT, raiseErrors,
+                    storeToUnknown, storeToGlobals);
+        }
+      }
+
+    } else if (CallExpr* call = toCallExpr(expr)) {
+      checkCall(fn, bb, call, symbols, OUT, raiseErrors,
+                storeToUnknown, storeToGlobals);
+
+    }
+
+    // Handle storeToUnknown, storeToGlobals
+    // TODO: storeToUnknown should only give up on
+    // other variables of compatible types
+    if (storeToGlobals || storeToUnknown) {
+      for_vector(Symbol, sym, symbols) {
+        if (sym->isConstValWillNotChange() == false) {
+          if (storeToUnknown || isOuterVar(sym, fn))
+            OUT[sym] = unknownAliasLocation();
+        }
+      }
+    }
+
+    if (trace) {
+      printAliasMap("+-", OUT, traceCopy);
+      delete traceCopy;
+    }
+  }
+}
+
+void findNilDereferences(FnSymbol* fn) {
+
+  bool debugging = 0 == strcmp(fn->name, debugNilsForFn) ||
+                   fn->id == debugNilsForId;
+
+  if (debugging) {
+    printf("Visiting function %s id %i\n", fn->name, fn->id);
+    nprint_view(fn);
+    gdbShouldBreakHere();
+  }
+
+  BasicBlock::buildBasicBlocks(fn);
+
+  // Gather the variables to consider
+  std::vector<Symbol*> idxToSym;
+  std::map<Symbol*, int> symToIdx;
+
+  {
+    int index = 0;
+    std::vector<SymExpr*> symExprs;
+    collectSymExprs(fn, symExprs);
+    for_vector(SymExpr, se, symExprs) {
+      Symbol* sym = se->symbol();
+      if (isArgSymbol(sym) || isVarSymbol(sym)) {
+        if (symToIdx.count(sym) == 0) {
+          // Add it to the map/array
+          symToIdx.insert(std::pair<Symbol*, int>(sym, index));
+          idxToSym.push_back(sym);
+          INT_ASSERT((int)(idxToSym.size()-1) == index);
+          index++;
+          if (debugging) {
+            printf("Tracking symbol %i %s\n", sym->id, sym->name);
+          }
+        }
+      }
+    }
+  }
```

Comment:
It will be really great to move the above block stmt to a separate function, ex. populateSymToIdx().

That will improve readability of findNilDereferences().